neps/nep-0491.md

```diff
@@ -0,0 +1,381 @@
+Status: Draft
+DiscussionsTo: https://gov.near.org/t/proposal-locking-account-storage-refunds-to-avoid-faucet-draining-attacks/34155
+Type: Protocol Track
+Version: 1.0.0
+Created: 2023-07-24
+LastUpdated: 2023-07-26
+---
+
+
+## Summary
+
+Non-refundable storage allows to create accounts with arbitrary state for users,
+without being susceptible to refund abuse.
+
+This is done by tracking non-refundable balance in a separate field of the
+account. This balance is only useful for storage staking and otherwise can be
+considered burned.
+
+
+## Motivation
+
+Creating new accounts on chain costs a gas fee and a storage staking fee. The
+more state is added to the account, the higher the storage staking fees. When
+deploying a contract on the account, it can quickly go above 1 NEAR per account.
+
+Some business models are okay with paying that fee for users upfront, just to
+get them onboarded. However, if a business does that today, their users can
+delete their new accounts and spend the tokens intended for storage staking in
+other ways. Since this is free for the user, they are financially incentivized
+to repeat this action for as long as the business has funds left in the faucet.
+
+The protocol should allow to create accounts in a way that is not susceptible to
+such refund abuse. This would at least change the incentives such that creating
+fake users is no longer profitable.
+
+Non-refundable storage staking is a further improvement over
+[NEP-448](https://github.com/near/NEPs/pull/448) (Zero Balance Accounts) which
+addressed the same issue but is limited to 770 bytes per account. By lifting the
+limit, sponsored accounts can be used in combination with smart contracts.
+
+## Specification
+
+Users can opt-in to nonrefundable storage when creating new accounts. For that,
+we use the new action `ReserveStorage`.
+
+```rust
+pub enum Action {
+    ...
+    ReserveStorage(ReserveStorageAction),
+    ...
+}
+```
+
+To create a named account today, the typical pattern is a transaction with
+`CreateAccount`, `Transfer`, and `AddKey`. To make the funds nonrefundable, we
+can use action `ReserveStorage` like this:
+
+```json
+"Actions": {
+  "CreateAccount": {},
+  "ReserveStorage": { "deposit": "1000000000000000000000000" },
+  "AddKey": { "public_key": "...", "access_key": "..." }
+}
+```
+
+Adding a `Transfer` action allows the combination of nonrefundable balance and
+refundable balance. This allows the user to make calls where they need to attach
+balance, for example an FT transfer which requires 1 yocto NEAR.
+
+```json
+"Actions": {
+  "CreateAccount": {},
+  "ReserveStorage": { "deposit": "1000000000000000000000000" },
+  "Transfer": { "deposit": "100" },
+  "AddKey": { "public_key": "...", "access_key": "..." }
+}
+```
+
+To create implicit accounts, the current protocol requires a single `Transfer`
+action without further actions in the same transaction and this has not changed
+with this proposal:
```

Comment:
I am confused here. It seems like it does change with this proposal?

Reply:
The creation of implicit accounts without reserved storage does not change, the request stays the same, I will remove the `ReserveStorage` action below